tasks/nginx.yml

```diff
@@ -0,0 +1,79 @@
+---
+- name: install nginx apt package
+  apt:
+    name: nginx
+
+- name: disable nginx default site
+  file:
+    path: /etc/nginx/sites-enabled/default
+    state: absent
+  notify:
+    - reload nginx
+
+- name: set up global nginx SSL settings
+  copy:
+    src: nginx-ssl
+    dest: /etc/nginx/conf.d/ssl.conf
+  notify:
+    - reload nginx
+
+- name: reload nginx if required
+  meta: flush_handlers
+
+- name: Set up Certbot
+  when: COMMON_SERVER_INSTALL_CERTBOT
+  block:
+
+    - name: copy nginx default configuration
+      template:
+        src: nginx-default-server.j2
+        dest: /etc/nginx/sites-available/common-server
+      notify:
+        - reload nginx
+
+    # We only copy the default configuration if we set up Certbot.  The default configuration makes
+    # nginx listen on port 80, which may not be desired in some cases (e.g. for the load balancer),
+    # but it is required if we want to reply to ACME challenges (which the load balancer does in a
+    # different way, configured in the HAProxy configuration).
+    - name: enable nginx site configuration
+      file:
+        src: /etc/nginx/sites-available/common-server
+        dest: /etc/nginx/sites-enabled/common-server
+        state: link
+      notify:
+        - reload nginx
+
+    - name: reload nginx if required
+      meta: flush_handlers
+
+    - name: open HTTP port on the firewall
+      ufw:
+        rule: allow
+        port: 80
+        proto: tcp
+
+    - name: add certbot PPA
+      apt_repository:
+        repo: ppa:certbot/certbot
+
+    - name: install certbot apt package
+      apt:
+        update_cache: yes
+        name: python-certbot-nginx
+
+    - name: Create Certbot webroot directory
+      file:
+        path: /var/www/certbot
+        state: directory
+
+    - name: generate SSL certificate with certbot
```

Comment:
Although this works, the job that the apt package installs is this:

```bash
# /etc/cron.d/certbot: crontab entries for the certbot package
#
# Upstream recommends attempting renewal twice a day
#
# Eventually, this will be an opportunity to validate certificates
# haven't been revoked, etc.  Renewal will only occur if expiration
# is within 30 days.
SHELL=/bin/sh
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin

0 */12 * * * root test -x /usr/bin/certbot -a \! -d /run/systemd/system && perl -e 'sleep int(rand(3600))' && certbot -q renew
```

I tried running that command with a dry run and it fails because:

```
Attempting to renew cert (rabbitmq-stage.opencraft.hosting) from /etc/letsencrypt/renewal/rabbitmq-stage.opencraft.hosting.conf produced an unexpected error: Problem binding to port 80: Could not bind to IPv4 or IPv6.. Skipping.
All renewal attempts failed. The following certs could not be renewed:
  /etc/letsencrypt/live/rabbitmq-stage.opencraft.hosting/fullchain.pem (failure)
```

So unless I did something wrong, I think we'd need to delete this default job and add one ourselves that uses the webroot plugin. The command that I used that worked in a dry run was the same that you used below. Using `renew` rather than `certonly` gives:

```
Currently, the renew verb is capable of either renewing all installed certificates that are due to be renewed or renewing a single certificate specified by its name. If you would like to renew specific certificates by their domains, use the certonly command instead. The renew verb may provide other options for selecting certificates to renew in the future.
```

so the command below is perfect in the cronjob.

Reply:
@UmanShahzad Certbot will reuse whatever auth plugin it used when the certificate was originally requested.  I may have to manually fix up the renewal configurations of existing certificates when I deploy this.

These servers could be affected:
* `chat.opencraft.com` – I already fixed that one manually.
* `rabbitmq*.opencraft.hosting` – these four servers need manual fixing.  I already fixed rabbitmq-stage in response to this comment, so `certbot renew --dry-run` now works.
* The accounting server should be fine without manual fixing.  It should have the old pre-hook and post-hook in its renewal config, so it should work correctly.  I can't verify this, though, so you may have to fix it up manually as well.

To fix this manually, we need to edit the configuration file in `/etc/letsencrypt/renewal/` to include these settings:
{code}
authenticator = webroot
webroot_path = /var/www/certbot,
[[webroot_map]]
chat.opencraft.com = /var/www/certbot
{code}
The last line needs to be adapted, of course.